contents/java/JavaConcurrency.md

```diff
@@ -0,0 +1,149 @@
+
+1. Thread A retrieves the value of counter and increments it to a value of 1
+2. Thread B then retrieves the values of counter, which is still 0, and decrements it to a value of -1
+3. Thread A stores the value of 1 to the counter variable
+4. Thread B stores the value of -1 to the counter variable and overwrites the previous value stored by Thread A
+
+Without thread interference, the expected value of the counter variable would be 0, since one thread increments it while the other decrements it. But with thread interference, the value of counter is simply the value written by the last thread. This is due to the unpredictable nature of threads as there is no way to know when the operating system switches between the two threads. 
+
+To solve this issue of interference, the keyword `synchronized` is used to ensure that method or block of code can only be accessed by one thread at a time. This is done through the use of the intrinsic lock system put in place by Java. Simply put, each object has it's own intrinsic lock which:
+* is acquired by the thread that works on it
+* prevents any other thread from working on it while it's lock has been acquired
+
+This can be illustrated by the following image where once Thread 2 (T2) acquires the lock for the synchronized block of code, the other two threads (T1 and T3) must wait for the synchronized block of code to release it's lock once T2 to complete its execution:
+
+![Synchronized block](http://www.javalearningacademy.com/wp-content/uploads/2015/10/Synchronized_Code_Block_And_Thread_Execution.jpg)
+
+For a deeper look, see the [Java Synchronisation](java/JavaSynchronization.md) section. 
+
+### Executors
+While it is easy to create one or two threads and run them, it becomes a problem when your application requires creating 20 or 30 threads for running tasks concurrently. This is where the `Executors` class comes in. It helps you with:
+
+1. **Thread Creation**<br>
+It provides various methods for creating threads, more specifically a pool of threads, uses to run tasks concurrently.
+
+1. **Thread Management**<br>
+It manages the life cycle of the threads in the thread pool. You don’t need to worry about whether the threads in the thread pool are active or busy or dead before submitting a task for execution.
+
+1. **Task submission and execution**<br>
+It provides methods for submitting tasks for execution in the thread pool, and also allows you to decide when the tasks will be executed. For example, a task can be submitted to be executed instantly, scheduled to be executed later or even executed periodically. 
+
+The `Executors` class provides convenient factory methods for creating the `ExecutorService` class. This class manages the lifecycle of tasks in various ways by assigning a `ThreadPool` to it. Different thread pools manage your tasks in different ways along with their own advantages and disadvantages. Some of these include:
```

Comment:
Is there a diagram to show how `Executors`, `ExecutorService` and `ThreadPool` are linked? :P